### PR TITLE
Upgrade to Pydantic v2

### LINKED ISSUE #1841: Upgrade to Pydantic v2
**Is your feature request related to a problem? Please describe.**
Pydantic v2 is the new default and offers better performance and new features. See also #1815 

**Describe the solution you'd like**
Upgrade Pydantic package

**Describe alternatives you've considered**
Update Pydantic and use `pydantic.v1` module for code that can't be easily migrated yet.

**Additional context**
See [migration guide](https://docs.pydantic.dev/latest/migration/) and also these points are relevant for upgrading specifically for this code base:
- Some field and method changes in `BaseModel`
- Model config changed (now a dictionary instead of a class)
- `@validator` and `@root_validator` are deprecated
- `pydantic.generics.GenericModel` is removed
- Changes in Required, optional, and nullable fields
- `TypeAdapter` may make root models unnecessary
- `BaseSettings` has moved to `pydantic-settings`

### PR DESCRIPTION
**Pull request by ammar92**
### Changes
This updates the Pydantic package to the latest major version. Please keep in mind that this should be tested properly:
- It should work on "old" data
- The OOI creation forms should still work
- Onboarding flow works as usual
- OOI detail and listing views work as usual

### Issue link
Closes #1841

---

### Code Checklist
- [x] All the commits in this PR are properly PGP-signed and verified;
- [x] This PR only contains functionality relevant to the issue; tickets have been created for newly discovered issues.
- [ ] I have written unit tests for the changes or fixes I made.
- [ ] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

### Communication
- [ ] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [ ] I have made corresponding changes to the documentation, if necessary.

---
## Checklist for code reviewers:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---
## Checklist for QA:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

**Comment by Darwinkel** [comment]
In-progress QA per @ammar92's request:

- No apparent bugs or regressions while browsing through KAT, even when CRUD'ing objects
- Some bits do not trigger:

image

File: boefjes/boefjes/config.py [hunk 1]
```diff
@@ -76,33 +79,38 @@ class Settings(BaseSettings):
...
+    def settings_customise_sources(
+        cls,
+        settings_cls: Type[BaseSettings],
+        init_settings: PydanticBaseSettingsSource,
+        env_settings: PydanticBaseSettingsSource,
+        dotenv_settings: PydanticBaseSettingsSource,
+        file_secret_settings: PydanticBaseSettingsSource,
+    ) -> Tuple[PydanticBaseSettingsSource, ...]:
+        backwards_compatible_settings = BackwardsCompatibleEnvSettings(settings_cls)
+        return env_settings, init_settings, file_secret_settings, backwards_compatible_settings
```
Comment (dekkers) [inline boefjes/boefjes/config.py:113]: Dotenv_settings is missing and the order is changed.

```suggestion
        return init_settings, env_settings,  dot_env_settings, file_secret_settings, backwards_compatible_settings
```

File: bytes/bytes/config.py [hunk 2]
```diff
@@ -128,34 +130,29 @@ class Settings(BaseSettings):
...
+    def settings_customise_sources(
+        cls,
+        settings_cls: Type[BaseSettings],
+        init_settings: PydanticBaseSettingsSource,
+        env_settings: PydanticBaseSettingsSource,
+        dotenv_settings: PydanticBaseSettingsSource,
+        file_secret_settings: PydanticBaseSettingsSource,
+    ) -> Tuple[PydanticBaseSettingsSource, ...]:
+        backwards_compatible_settings = BackwardsCompatibleEnvSettings(settings_cls)
+        return env_settings, init_settings, file_secret_settings, backwards_compatible_settings
```
Comment (dekkers) [inline bytes/bytes/config.py:148]: Dotenv_settings is missing and the order is changed.

```suggestion
        return init_settings, env_settings,  dot_env_settings, file_secret_settings, backwards_compatible_settings
```

File: octopoes/octopoes/config/settings.py [hunk 3]
```diff
@@ -55,40 +55,45 @@ class Settings(BaseSettings):
...
+    def settings_customise_sources(
+        cls,
+        settings_cls: Type[BaseSettings],
+        init_settings: PydanticBaseSettingsSource,
+        env_settings: PydanticBaseSettingsSource,
+        dotenv_settings: PydanticBaseSettingsSource,
+        file_secret_settings: PydanticBaseSettingsSource,
+    ) -> Tuple[PydanticBaseSettingsSource, ...]:
+        backwards_compatible_settings = BackwardsCompatibleEnvSettings(settings_cls)
+        return env_settings, init_settings, file_secret_settings, backwards_compatible_settings
```
Comment (dekkers) [inline octopoes/octopoes/config/settings.py:96]: Dotenv_settings is missing and the order is changed.

```suggestion
        return init_settings, env_settings,  dot_env_settings, file_secret_settings, backwards_compatible_settings
```

File: bytes/bytes/models.py [hunk 4]
```diff
@@ -28,17 +29,14 @@ class HashingRepositoryReference(str, Enum):
...
+    return parsed
 
-    @classmethod
-    def validate(cls, value: Union[datetime, StrBytesIntFloat]) -> datetime:
-        parsed = parse_datetime(value)
-        if parsed.tzinfo is None or parsed.tzinfo.utcoffset(parsed) is None:
-            raise ValueError(f"{parsed} is not timezone aware")
-        return parsed
+
+TimezoneAwareDatetime = Annotated[datetime, AfterValidator(_validate_timezone_aware_datetime)]
```
Comment (dekkers) [inline bytes/bytes/models.py:39]: Pydantic V2 has the `AwareDatetime` type that we can use.

**Comment by Rieven** [comment]
After checking out to this branch and a `make kat` I still had some old OOIs which a good way to test backwards compatibility issues. 
***
OOI detail Tree and Graph tabs
***

Go to object detail page and go to the "Tree" and "Graph" tab. Error found there: THIS IS ONLY FOR OLD OOIs

Request Method: GET
Request URL: http://localhost:8000/en/_steven/objects/graph/?ooi_id=DKIMExists%7Cinternet%7Crieven.com
`Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/django/views/generic/base.py", line 143, in dispatch
    return handler(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/rocky/rocky/views/ooi_view.py", line 83, in get
    return super().get(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/django/views/generic/base.py", line 226, in get
    context = self.get_context_data(**kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/rocky/rocky/views/ooi_tree.py", line 90, in get_context_data
    context = super().get_context_data(**kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/rocky/rocky/views/ooi_tree.py", line 54, in get_context_data
    context["tree"] = self.get_filtered_tree(self.get_tree_dict())
                                             ^^^^^^^^^^^^^^^^^^^^
  File "/app/rocky/rocky/views/ooi_tree.py", line 21, in get_tree_dict
    return create_object_tree_item_from_ref(self.tree.root, self.tree.store)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/rocky/tools/ooi_helpers.py", line 160, in create_object_tree_item_from_ref
    create_object_tree_item_from_ref(child, ooi_store, knowledge_base, depth, child_position, location)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/rocky/tools/ooi_helpers.py", line 160, in create_object_tree_item_from_ref
    create_object_tree_item_from_ref(child, ooi_store, knowledge_base, depth, child_position, location)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/rocky/tools/ooi_helpers.py", line 160, in create_object_tree_item_from_ref
    create_object_tree_item_from_ref(child, ooi_store, knowledge_base, depth, child_position, location)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/rocky/tools/ooi_helpers.py", line 140, in create_object_tree_item_from_ref
    ooi = ooi_store[str(reference_node.reference)]
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Exception Type: KeyError at /en/_steven/objects/graph/
Exception Value: 'URL|internet|https://rieven.com'
***
HEALTH PAGE
***
Request Method: GET
Request URL: http://localhost:8000/en/_rieven/health/v1/

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/django/views/generic/base.py", line 143, in dispatch
    return handler(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/django/views/generic/base.py", line 226, in get
    context = self.get_context_data(**kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/rocky/rocky/views/health.py", line 123, in get_context_data
    rocky_health = get_rocky_health(self.octopoes_api_connector)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/rocky/rocky/views/health.py", line 93, in get_rocky_health
    rocky_health = ServiceHealth(
                   
  File "/usr/local/lib/python3.11/site-packages/pydantic/main.py", line 164, in __init__
    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Exception Type: ValidationError at /en/_rieven/health/v1/
Exception Value: 1 validation error for ServiceHealth
results.0
  Input should be a valid dictionary or instance of ServiceHealth [type=model_type, input_value=ServiceHealth(service='oc...12766614}, results=[])]), input_type=ServiceHealth]
    For further information visit https://errors.pydantic.dev/2.4/v/model_type

**Comment by ammar92** [comment]
Thanks for testing out @Rieven Is there any way you could also share the local variables dump for each stack trace? In Django you can expand the local vars; this would make it easier to debug some of these `KeyError` and `ValidationError` exceptions

**Comment by Rieven** [comment]
> Thanks for testing out @Rieven Could you also share the local variables dump for each stack trace? In Django you can expand the local vars; this would make it easier to debug some of these `KeyError` and `ValidationError` exceptions

[KeyError at _en__steven_objects_tree_.pdf](https://github.com/minvws/nl-kat-coordination/files/13204353/KeyError.at._en__steven_objects_tree_.pdf)

[ValidationError at _en__steven_health_v1_.pdf](https://github.com/minvws/nl-kat-coordination/files/13204368/ValidationError.at._en__steven_health_v1_.pdf)

**Review by dekkers — APPROVED** [review]

**Pull request merged**